pkg/server/plugin/nodeattestor/k8s/k8s.go

```diff
@@ -0,0 +1,210 @@
+package k8s
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/plugin/k8s"
+	"github.com/spiffe/spire/pkg/common/plugin/x509pop"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/common"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+	"github.com/spiffe/spire/proto/server/nodeattestor"
+)
+
+const (
+	pluginName = "k8s"
+)
+
+type configuration struct {
+	trustDomain string
+	trustBundle *x509.CertPool
+}
+
+type K8sConfig struct {
+	TrustDomain  string `hcl:"trust_domain"`
+	CABundlePath string `hcl:"ca_bundle_path"`
+}
+
+type K8sPlugin struct {
+	m sync.Mutex
+	c *configuration
+}
+
+func newError(format string, args ...interface{}) error {
+	return fmt.Errorf("k8s: "+format, args...)
+}
+
+func buildSelectors(leaf *x509.Certificate, chains [][]*x509.Certificate) []*common.Selector {
+	selectors := []*common.Selector{}
+
+	if leaf.Subject.CommonName != "" {
+		selectors = append(selectors, &common.Selector{
+			Type: "k8s", Value: "subject:cn:" + leaf.Subject.CommonName,
+		})
+	}
+
+	// Used to avoid duplicating selectors.
+	fingerprints := map[string]*x509.Certificate{}
+	for _, chain := range chains {
+		// Iterate over all the certs in the chain (skip leaf at the 0 index)
+		for _, cert := range chain[1:] {
+			fp := x509pop.Fingerprint(cert)
+			// If the same fingerprint is generated, continue with the next certificate, because
+			// a selector should have been already created for it.
+			if _, ok := fingerprints[fp]; ok {
+				continue
+			}
+			fingerprints[fp] = cert
+
+			selectors = append(selectors, &common.Selector{
+				Type: "k8s", Value: "ca:fingerprint:" + fp,
+			})
+		}
+	}
+
+	return selectors
+}
+
+func New() *K8sPlugin {
+	return &K8sPlugin{}
+}
+
+func (p *K8sPlugin) Attest(stream nodeattestor.Attest_PluginStream) error {
+	req, err := stream.Recv()
+	if err != nil {
+		return err
+	}
+
+	c := p.getConfiguration()
+	if c == nil {
+		return newError("not configured")
+	}
+
+	if dataType := req.AttestationData.Type; dataType != pluginName {
+		return newError("unexpected attestation data type %q", dataType)
+	}
+
+	attestationData := new(x509pop.AttestationData)
+	if err := json.Unmarshal(req.AttestationData.Data, attestationData); err != nil {
```

Comment:
Same comment here - what if `req.AttestationData` is nil?

Reply:
will fix